### PR TITLE
Adding some quality of life for `pipeline` function.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Adds two quality of life things on `pipeline` function.

- Make `task` optional `pipeline(model="gpt2")` should not be enough (will rely on the hub).
- Make `pipeline_class` overridable to ease override for users of the `pipeline` function.
  `pipeline(model="gpt2", pipeline_class=MyPipeline)` should enable you to use whatever
  pipeline you want.

Related to #14278

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LysandreJik

### CODE REVIEW

File: src/transformers/pipelines/__init__.py [hunk 1]
```diff
@@ -248,6 +251,29 @@
...
+    try:
+        http_get(f"https://huggingface.co/api/models/{model}", tmp, headers=headers)
+        tmp.seek(0)
+        body = tmp.read()
+        data = json.loads(body)
+    except Exception as e:
+        raise RuntimeError(f"Instantiating a pipeline without a task set raised an error: {e}")
+    if "pipeline_tag" not in data:
+        raise RuntimeError(
+            f"The model {model} does not seem to have a correct `pipline_tag` set to infer the task automatically"
```
Comment (sgugger) [inline src/transformers/pipelines/__init__.py:269]:
```suggestion
            f"The model {model} does not seem to have a correct `pipeline_tag` set to infer the task automatically"
```

File: tests/test_pipelines_common.py [hunk 2]
```diff
@@ -261,6 +263,27 @@ def __getitem__(self, i):
         for output in text_classifier(dataset):
             self.assertEqual(output, {"label": ANY(str), "score": ANY(float)})
 
+    @require_torch
+    def test_check_task_auto_inference(self):
+        pipeline(model="Narsil/tiny-distilbert-sequence-classification")
+
```
Comment (sgugger) [inline tests/test_pipelines_common.py:269]: Can we check that the result has the right type?

**Review by sgugger — APPROVED** [review]
Those look like great improvements, thanks for adding support!

File: tests/test_pipelines_common.py (hunk 2 above)
Comment (Narsil) [reply tests/test_pipelines_common.py:271]: Done it, thanks !

**Comment by Narsil** [comment]
I'll merge ahead.

**Pull request merged**